src/verify_cert.rs

```diff
@@ -217,14 +217,32 @@ fn check_signatures(
 }
 
 fn check_crl(
-    _supported_sig_algs: &[&SignatureAlgorithm],
+    supported_sig_algs: &[&SignatureAlgorithm],
     cert: &Cert,
-    _issuer_subject: untrusted::Input,
-    _issuer_spki: untrusted::Input,
+    issuer_subject: untrusted::Input,
+    issuer_spki: untrusted::Input,
     crl_provider: &dyn CrlProvider,
 ) -> Result<(), Error> {
-    let _relevant_crl = crl_provider.crl_for_cert(cert);
-    todo!();
+    assert_eq!(cert.issuer, issuer_subject);
+
+    let relevant_crl = crl_provider.crl_for_cert(cert);
+    match relevant_crl {
+        None => Ok(()),
+        Some(crl) => {
+            // Verify the CRL signature with the issuer SPKI.
+            signed_data::verify_signed_data(supported_sig_algs, issuer_spki, &crl.signed_data)?;
```

Comment:
After paging this work back into my brain I'm remembering some of why this ends up being challenging. 

I think from a design perspective we don't want to push the signature validation to the user's `CrlProvider` implementation because it will be too easy to skip doing it, or doing it incorrectly so in my mind that option can be ruled out easily.

We wouldn't be able to do this with the `CertificateRevocationList::TryFrom` impl (since we wouldn't be able to provide an issuer argument in addition to the DER while meeting the trait's signature) so we would have to remove that trait to make progress on this front.

We could provide our own constructor that allowed specifying the issuer to use for validation, but we'd run into a complication from 5280. [RFC 5280 §6.3.3](https://www.rfc-editor.org/rfc/rfc5280#section-6.3.3) says:
>  (f)  Obtain and validate the certification path for the issuer of
>       the complete CRL.  The trust anchor for the certification
>       path MUST be the same as the trust anchor used to validate
>       the target certificate.

The MUST there will be tricky to enforce at CRL-load time when we won't know the trust anchor that will be used as the root for path-building until we actually build the path with the bundle of intermediates and trust anchors provided to `build_chain`. 

It also seems appealing to me that you could load a bunch of CRLs for end entities and intermediates without needing to source the intermediate issuers for validation ahead of time. If the CRL is relevant based on the intermediates used during path-building (likely sourced from the TLS handshake) they'll be used as required.

What do you think? Is there a middle ground I'm missing? :thinking:

Reply:
If we do end up wanting to do CRL signature validation at the time-of-load another challenge will be extracting and generalize the existing `build_chain` code to offer a way to verify a chain of signatures to a trust root without having an end entity certificate in-hand. 

The existing code is tightly coupled to the `Cert` representation and the `EndEntityOrCa` type that's used to build a linked list. We'd want to be able to express something like "CrlOrCa" instead. I don't love the idea of having two path building implementations to maintain, but I also struggled to generalize the existing one without making a mess. That also motivated my decision to consider revocation + the CRL signature validation as part of the existing `check_singatures` fn that's already walking the certificate linked list.

Reply:
Not to distract from the consensus that we should postpone the optimization here, but it sounds like we should store the CRL and trust anchor together? That way we could assert in the type system that we've verified that the trust anchor has signed the CRL.

Reply:
> - perhaps we should just remember this as a future optimisation?

That sounds reasonable. I'll file an issue and leave a comment to point to it.

> it sounds like we should store the CRL and trust anchor together? That way we could assert in the type system that we've verified that the trust anchor has signed the CRL.

@djc Just to make sure I understand: you're suggesting that as a way to address the 5280 MUST concern in a world where we do the refactoring suggested by ctz, or is it a change you think should be incorporated with the current code?

Reply:
The former.

Reply:
:+1: Noted that on the tracking issue so it won't get lost when I resolve this comment: https://github.com/rustls/webpki/issues/81#issuecomment-1591587529